pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/ServerRequestPlanVisitor.java

```diff
@@ -0,0 +1,363 @@
+    if (dataSource != null) {
+      QuerySource querySource = new QuerySource();
+      querySource.setTableName(dataSource.getTableName());
+      brokerRequest.setQuerySource(querySource);
+    }
+
+    // 3. create instance request with segmentList
+    InstanceRequest instanceRequest = new InstanceRequest();
+    instanceRequest.setRequestId(requestId);
+    instanceRequest.setBrokerId("unknown");
+    instanceRequest.setEnableTrace(false);
+    instanceRequest.setSearchSegments(segmentList);
+    instanceRequest.setQuery(brokerRequest);
+
+    context.setInstanceRequest(instanceRequest);
+    return context;
+  }
+
+
+
+
+  private static void walkStageNode(StageNode node, ServerPlanRequestContext context) {
+    node.visit(getInstance(), context);
+  }
+
+  @Override
+  public Void visitAggregate(AggregateNode node, ServerPlanRequestContext context) {
+    visitChildren(node, context);
+    // set group-by list
+    context.getPinotQuery().setGroupByList(CalciteRexExpressionParser.convertGroupByList(
+        node.getGroupSet(), context.getPinotQuery()));
+    // set agg list
+    context.getPinotQuery().setSelectList(CalciteRexExpressionParser.addSelectList(
+        context.getPinotQuery().getGroupByList(), node.getAggCalls(), context.getPinotQuery()));
+    return _aVoid;
+  }
+
+  @Override
+  public Void visitFilter(FilterNode node, ServerPlanRequestContext context) {
+    visitChildren(node, context);
+    context.getPinotQuery().setFilterExpression(CalciteRexExpressionParser.toExpression(
+        node.getCondition(), context.getPinotQuery()));
+    return _aVoid;
+  }
+
+  @Override
+  public Void visitJoin(JoinNode node, ServerPlanRequestContext context) {
+    // visit only the static side.
+    StageNode staticSide = node.getInputs().get(0);
+    StageNode dynamicSide = node.getInputs().get(1);
+    if (staticSide instanceof MailboxReceiveNode) {
+      dynamicSide = node.getInputs().get(0);
+      staticSide = node.getInputs().get(1);
+    }
+    staticSide.visit(this, context);
+
+    // rewrite context for JOIN as dynamic filter
+    // step 1: get the other-side's result.
+    MailboxReceiveNode dynamicMailbox = (MailboxReceiveNode) dynamicSide;
+    List<ServerInstance> sendingInstances = context.getMetadataMap().get(dynamicMailbox.getSenderStageId())
+        .getServerInstances();
+    MailboxReceiveOperator mailboxReceiveOperator = new MailboxReceiveOperator(context.getMailboxService(),
+        dynamicMailbox.getDataSchema(), sendingInstances, dynamicMailbox.getExchangeType(),
+        dynamicMailbox.getPartitionKeySelector(), context.getHostName(), context.getPort(),
+        context.getRequestId(), dynamicMailbox.getSenderStageId());
+    TransferableBlock block = attachDynamicOperatorResults(mailboxReceiveOperator, dynamicMailbox.getDataSchema());
+    context.setDynamicOperatorResult(block);
+    // step 2: write filter expression
+    attachDynamicFilter(context.getPinotQuery(), node.getJoinKeys(), block);
```

Comment:
This is the fun part. we run the mailbox receive here and generate the IDSet(s) for equality join conditions, rewrite them into filter clause.

Reply:
😆 I should've just read your comment first before I spent a lot of time confirming my theory that this doesn't work if there's any select columns on the left side.

Reply:
:-) the only reason why it doesnt work is b/c it doesn't conform with the expected dataSchema in the receiving end of the mailbox. but the data is there. just not glue together